crates/subspace-farmer/src/plot.rs

```diff
@@ -443,37 +443,55 @@ impl WeakPlot {
     }
 }
 
+/// Mapping from piece index hash to piece offset.
+///
+/// Piece index hashes are transformed in the following manner:
+/// - Assume that farmer address is the middle (`2 ^ 255`) of the `PieceDistance` field
+/// - Move every piece according to that
 #[derive(Debug)]
 struct IndexHashToOffsetDB {
     inner: DB,
     address: PublicKey,
-    max_distance: Option<PieceDistance>,
+    /// The first one is the lower bound of plot, the second one is the upper bound
+    key_bounds: Option<(PieceDistance, PieceDistance)>,
 }
 
 impl IndexHashToOffsetDB {
-    fn update_max_distance(&mut self) {
-        self.max_distance = {
+    fn update_key_bounds(&mut self) {
+        self.key_bounds = try {
             let mut iter = self.inner.raw_iterator();
+            iter.seek_to_first();
+            let lower_bound = iter.key().map(PieceDistance::from_big_endian)?;
             iter.seek_to_last();
-            iter.key().map(PieceDistance::from_big_endian)
+            let upper_bound = iter.key().map(PieceDistance::from_big_endian)?;
+            (lower_bound, upper_bound)
         };
     }
 
+    fn get_key(&self, index_hash: &PieceIndexHash) -> PieceDistance {
+        let address = PieceDistance::from_big_endian(self.address.as_ref());
+        let index_hash = PieceDistance::from_big_endian(&index_hash.0);
+
+        // We do not care about overflows here
+        (PieceDistance::MAX / 2u64)
+            .overflowing_add(index_hash.overflowing_sub(address).0)
+            .0
```

Comment:
Honestly I don't quite follow the logic in conjunction with the comment. Obviously we care about overflow, otherwise you wouldn't use `overflowing_*` functions. I believe you want to use `wrapping_*` functions instead (which should also map better to machine instructions)?

Reply:
Actually I remembered that `uint` crate doesn't implement `wrapping_*` methods, that's why I used `overflowing_*` here. Will manually implement `wrapping_sub`

Reply:
Would be nice to add it upstream, can you open PR there?

Reply:
I guess they won't take it. They do not implement `num` traits anyways and having `wrapping_*` methods seem to be pointless as there is `overflowing_*`. Knowing upfront that you won't need overflow flag doesn't bring here performance benefits.

Reply:
They are not quite the same and in some cases makes [significant difference](https://godbolt.org/#g:!((g:!((g:!((h:codeEditor,i:(filename:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,selection:(endColumn:19,endLineNumber:1,positionColumn:19,positionLineNumber:1,selectionStartColumn:19,selectionStartLineNumber:1,startColumn:19,startLineNumber:1),source:'pub+fn+overflowing_add(a:+i32,+b:+i32)+-%3E+i32+%7B%0A++++a.overflowing_add(b).0%0A%7D%0A%0Apub+fn+wrapping_add(a:+i32,+b:+i32)+-%3E+i32+%7B%0A++++a.wrapping_add(b)%0A%7D'),l:'5',n:'0',o:'Rust+source+%231',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0'),(g:!((h:compiler,i:(compiler:r1600,filters:(b:'0',binary:'1',commentOnly:'0',demangle:'0',directives:'0',execute:'1',intel:'0',libraryCode:'0',trim:'1'),flagsViewOpen:'1',fontScale:14,fontUsePx:'0',j:1,lang:rust,libs:!(),options:'',selection:(endColumn:1,endLineNumber:1,positionColumn:1,positionLineNumber:1,selectionStartColumn:1,selectionStartLineNumber:1,startColumn:1,startLineNumber:1),source:1,tree:'1'),l:'5',n:'0',o:'rustc+1.60.0+(Rust,+Editor+%231,+Compiler+%231)',t:'0')),k:50,l:'4',n:'0',o:'',s:0,t:'0')),l:'2',n:'0',o:'',t:'0')),version:4), though with `-C opt-level=3` it does compile to the same code. Not sure it is as cleanly optimized in `uint` case, likely so. Either way what we care about here is wrapping computation specifically, so if possible code should point that out by using corresponding method.

Reply:
Ikr, but here you actually multiply with overflow 8 big words with each other, so I bet checking overflow is pretty minor thing here.